packages/alfa-css/src/value/shape/rectangle.ts

```diff
@@ -0,0 +1,150 @@
+  }
+
+  public get format(): "rectangle" {
+    return "rectangle";
+  }
+
+  public get top(): Length | Keyword<"auto"> {
+    return this._top;
+  }
+
+  public get right(): Length | Keyword<"auto"> {
+    return this._right;
+  }
+
+  public get bottom(): Length | Keyword<"auto"> {
+    return this._bottom;
+  }
+
+  public get left(): Length | Keyword<"auto"> {
+    return this._left;
+  }
+
+  public equals(value: unknown): value is this {
+    return (
+      value instanceof Rectangle &&
+      value._top.equals(this._top) &&
+      value._right.equals(this._right) &&
+      value._bottom.equals(this._bottom) &&
+      value._left.equals(this._left)
+    );
+  }
+
+  public hash(hash: Hash) {
+    this._top.hash(hash);
+    this._right.hash(hash);
+    this._bottom.hash(hash);
+    this._left.hash(hash);
+  }
+
+  public toJSON(): Rectangle.JSON {
+    return {
+      type: "shape",
+      format: "rectangle",
+      top: this._top.toJSON(),
+      right: this._right.toJSON(),
+      bottom: this._bottom.toJSON(),
+      left: this._left.toJSON(),
+    };
+  }
+
+  public toString(): string {
+    return `rect(${this._top.toString()}, ${this._right.toString()}, ${this._bottom.toString()}, ${this._left.toString()})`;
+  }
+}
+
+export namespace Rectangle {
+  export interface JSON extends Value.JSON {
+    type: "shape";
+    format: "rectangle";
+    top: Length.JSON | Keyword.JSON;
+    right: Length.JSON | Keyword.JSON;
+    bottom: Length.JSON | Keyword.JSON;
+    left: Length.JSON | Keyword.JSON;
+  }
+
+  export function isRectangle(value: unknown): value is Rectangle {
+    return value instanceof Rectangle;
+  }
+
+  const parseLengthAuto = (input: Slice<Token>) => {
+    return either(Length.parse, Keyword.parse("auto"))(input);
+  };
+
+  export const parse: Parser<Slice<Token>, Rectangle, string> = (input) =>
+    right(
+      Token.parseFunction((fn) => fn.value === "rect"),
+      left(
+        either(
+          // If there is the wrong separator, separatedList still return the first value.
+          // Thus, it doesn't fail and the full parser fails at the closing parenthesis.
+          // We need to peek to find the correct separator…
+          right(
+            peek(pair(parseLengthAuto, Token.parseWhitespace)),
+            separatedList(parseLengthAuto, Token.parseWhitespace)
+          ),
+          separatedList(
+            parseLengthAuto,
+            pair(Token.parseComma, option(Token.parseWhitespace))
+          )
+        ),
+        Token.parseCloseParenthesis
+      )
+    )(input).flatMap(([remainder, result]) => {
+      const values = [...result];
+
+      const err: Result<[Slice<Token>, Rectangle], string> = Err.of(
+        "rect() must have exactly 4 arguments"
+      );
+
+      if (values.length !== 4) {
+        return err;
+      }
+      return Ok.of<[Slice<Token>, Rectangle]>([
+        remainder,
+        Rectangle.of(values[0], values[1], values[2], values[3]),
+      ]);
+    });
+}
```

Comment:
I've got a draft of the different shapes sitting in my local tree as I was experimenting a little while back, based on CSS Shapes: https://drafts.csswg.org/css-shapes/#supported-basic-shapes. I'll get them wrapped up and pushed.

Reply:
Perfect, implementing these would have been my next step 😀

Reply:
The `rect()` shape corresponds to the new `inset()` shape from CSS Shapes:

https://github.com/Siteimprove/alfa/blob/18c3beb1d09bccb615b9d24c7cfbd1d09992ae87/packages/alfa-css/src/value/shape/inset.ts#L4-L36

Reply:
Looking more into it, I think that `rect()` and `inset()` are sufficiently different that two separate implementations will be easier. Notably:
* `inset()` arguments are `<length-percentage>`, `rect()` are `<length>` or `auto`. So, type `O` above would be different in both, even inheriting (to fix the radius at 0) would prove tricky.
* parse-wise, they are fairly different also, `inset()` accept less than 4 args (same meaning as for `border`) while `rect()` requires 4 args.

I think it's just easier to treat them as different shapes rather than trying to force `rect()` as being an `inset()`.
=> I think this PR is actually OK to merge as is and #591 keeps moving 😄

Reply:
Apologies for stagnating on this! I think you're right that the `rect()` shape warrants its own separate definition, though then at the very least it should be added in #591 as a deprecated basic shape that only allows the `border-box` box value in `Shape`.

Reply:
I feel like I'm missing a connection 😅 
`rect()` can only be used in `clip`, and `clip` can only use `rect()` and `auto`. I don't see how `rect()` interacts with the various `<*-box>` types.

But then yes, putting an `export type Deprecated = Rectangle` in the `Shape` namespace is more or less part of the plan 😄

Reply:
I added the basics so there are still some blanks to fill in: https://github.com/Siteimprove/alfa/blob/6578b25bcacb3d5aa982b5980e960b8b48923f45/packages/alfa-css/src/value/shape/rectangle.ts#L4-L24